How to check internal transactions (#389)
matter-labs/zksync-web-era-docs: docs/dev/troubleshooting/withdrawal-delay.md

- Delayed execution affects not only the standard zkSync ETH and ERC20 bridges but also any custom bridge built by a different team.
- Implementing the logic in an external governor-controlled contract makes it easy to remove this limitation later.

## Why can't I find my transaction in the block explorer?

There are several reasons why your withdrawal may be successful but you cannot see your tokens, or the transaction, on the block explorer. For example:

1. **Delay in block confirmation**: as mentioned in the [withdrawal delay](#withdrawal-delay) section above, successful withdrawals may be subject to an ongoing confirmation process. Consequently, you won't be able to see your transaction on the block explorer until the confirmation process completes.

2. **Transaction reverted**: a withdrawal may fail to appear on the block explorer if the transaction is reverted due to a conflict or issue with the smart contract that executed the transaction. In such cases, the transaction is canceled and it won't be included on the block explorer.

> 3. **Wrong address**: If a user mistakenly sends tokens to an incorrect address, the transaction will not be visible on the block explorer, and the tokens will not arrive in the intended recipient's wallet.

Reviewer: ```suggestion
3. **Wrong address**: if a user mistakenly sends tokens to an incorrect address, the transaction will not be visible on the block explorer, and the tokens will not arrive in the intended recipient's wallet.
```

On the other hand, **Internal Transactions** are initiated by a smart contract or other internal code execution within the Ethereum network. Although these transactions can be prompted by user activity, they are not sent directly from one address to another but are instead part of the internal workings of a smart contract. Internal transactions may involve the transfer of ETH or ERC20 tokens between different addresses within the contract. 

::: tip
Withdrawals from the zkSync Era network are typically internal transactions managed by the [zkSync Era Diamond Proxy](https://etherscan.io/address/0x32400084c286cf3e17e7b677ea9583e60a000324) contract. These transactions are recorded in the **Internal Transactions** section in Etherscan due to their internal nature.
:::

In summary, the **Transactions** section in Etherscan displays transactions between external addresses, while the **Internal Transactions** section displays transactions that occur within smart contracts.

### How to check your internal transactions

> To check the internal transactions of an Ethereum wallet, follow these steps:

Reviewer: ```suggestion
```

Reviewer: The title of the section says what's going to happen in the section, you don't need to repeat yourself.

1. Launch your web browser and visit [Etherscan.io](https://etherscan.io/).

2. In the search bar at the top of the page, enter the Ethereum wallet address you used to withdraw funds. This should be the address you withdrew funds from, not the destination address.

![Etherscan search bar](../../assets/images/ether-search.png "Search for transactions")

3. Click **Enter**.

4. Scroll down to the **Internal Transactions** section.

> Refer to the image below for more clarity.

Reviewer: ```suggestion
```

Reviewer: This is unnecessary, unless there is something weird about the image that means the user wouldn't look at it automatically.

# Withdrawal delay

> In order to prevent a quick drain of the protocol in case a critical bug has been discovered and exploited, we are introducing a block execution delay. Each L2 block committed to L1 will have a time lock before it is executed and finalized. This means that there is enough time to verify the effects of the transactions included in a block before the block becomes final. The zkSync team will be monitoring each block and be able to investigate any anomaly (e.g. rapid outflow, unusually large withdrawals, etc).

Reviewer: ```suggestion
In order to prevent a quick drain of the protocol in the case a critical bug has been discovered and exploited, we are introducing a block execution delay. Each L2 block committed to L1 will have a time lock before it is executed and finalized. This means that there is enough time to verify the effects of the transactions included in a block before the block becomes final. The zkSync team will be monitoring each block and investigating any anomaly (e.g. rapid outflow, unusually large withdrawals, etc).
```

> To introduce this time lock no changes were made to the audited smart contracts. Instead, we have used an existing Validator role that we control and additionally restricted ourselves by pointing it at an intermediate smart contract with a time lock. It will be initially configured for a **24-hour** delay, which will be gradually decreased as the system matures. Changing the delay requires multiple signatures collected from several cold wallets owned by zkSync leadership.

Reviewer: ```suggestion
To introduce this time lock, no changes were made to the audited smart contracts. Instead, we have used an existing Validator role that we control and, additionally, that we restricted by pointing it to an intermediate smart contract with a time lock. The time lock is initially configured for a **24-hour** delay, which will gradually decrease as the system matures. Changing the delay requires multiple signatures collected from several cold wallets owned by zkSync leadership.
```

This design has the following advantages:

- Even if an attacker finds a critical bug in ZK circuits and also successfully compromises the servers running our sequencer, there is plenty of time to detect an exploit, investigate, and freeze the protocol via governance.
- No changes were introduced to the zkSync Era contracts, so even if the intermediate contract is compromised we revert back to the original state.
- Delayed execution affects not only the standard zkSync ETH and ERC20 bridges but also any custom bridge built by a different team.
- Implementing the logic in an external governor-controlled contract makes it easy to remove this limitation later.

> 
> ## Why can't i find my transaction in the block explorer?
> 
> There could be several reasons why your withdrawal was completed, but you cannot see your tokens or the transaction on the block explorer. Some of these reasons include:
> 
> 1. Delay in Block Confirmation: As previously indicated, a successfully executed withdrawal transaction may undergo an ongoing confirmation process, which requires a certain amount of time before the block containing the transaction is confirmed. Consequently, the transaction cannot be observed on the block explorer until the confirmation process is completed.

Reviewer: ```suggestion

## Why can't I find my transaction in the block explorer?

There are several reasons why your withdrawal may have completed but you cannot see your tokens or the transaction on the block explorer. For example:

1. **Delay in block confirmation**: as mentioned in the [withdrawal delay](#withdrawal-delay) section above, successfully withdrawals may be subject to an ongoing confirmation process. Consequently, you won't be able to see your transaction in the block explorer until the confirmation process completes.
```

> 
> 2. Transaction Reverted: A withdrawal may fail to appear on the block explorer if the transaction is reverted due to a conflict or issue with the smart contract that executes the transaction. In such cases, the transaction is canceled, rendering it invisible to users on the block explorer.

Reviewer: ```suggestion

2. **Transaction reverted**: a withdrawal may fail to appear on the block explorer if the transaction is reverted due to a conflict or issue with the smart contract that executed the transaction. In such cases, the transaction is canceled and it won't be included on the block explorer.
```

> 3. Wrong Address: If a user mistakenly sends tokens to an incorrect address, the transaction will not be visible on the block explorer, and the tokens will not reflect in the user's wallet.
> 

Reviewer: ```suggestion
3. **Wrong address**: If a user mistakenly sends tokens to an incorrect address, the transaction will not be visible on the block explorer, and the tokens will not arrive in the intended recipient's wallet.

```

> ### Transactions in etherscan
> 
> The **Transactions** section of Etherscan displays transactions that occur between two Ethereum addresses. It shows details such as the transaction hash, the block number, the timestamp, the sender and receiver addresses, the amount of Ether or tokens involved, and the transaction fee. These transactions are confirmed by the network and are typically sent from one external address to another.
> 
> On the other hand, **Internal Transactions** are initiated by a smart contract or other internal code execution within the Ethereum network. Although these transactions can be prompted by a user action, they are not sent directly from one address to another but are instead part of the internal workings of a smart contract. Internal transactions may involve the transfer of ETH or ERC-20 tokens between different addresses within the contract. 

Reviewer: ```suggestion
### Transactions in Etherscan

The **Transactions** section in Etherscan displays transactions between two Ethereum addresses. It shows details such as the transaction hash, the block number, the timestamp, the sender and receiver addresses, the amount of Ether or tokens involved, and the transaction fee. These transactions are confirmed by the network and are typically sent from one external address to another.

On the other hand, **Internal Transactions** are initiated by a smart contract or other internal code execution within the Ethereum network. Although these transactions can be prompted by user activity, they are not sent directly from one address to another but are instead part of the internal workings of a smart contract. Internal transactions may involve the transfer of ETH or ERC-20 tokens between different addresses within the contract. 
```

> 
> ::: tip
> 
> Withdrawals from the zkSync Era network are typically internal transactions managed by the [zkSync Era Diamond Proxy](https://etherscan.io/address/0x32400084c286cf3e17e7b677ea9583e60a000324) contract. These transactions are recorded in the **"Internal Transactions"** section of Etherscan due to their internal nature.
> 
> :::
> 
> 
> In summary, the **Transactions** section of Etherscan displays transactions between external addresses, while the **Internal Transactions** section displays transactions that occur within smart contracts.

Reviewer: ```suggestion

::: tip
Withdrawals from the zkSync Era network are typically internal transactions managed by the [zkSync Era Diamond Proxy](https://etherscan.io/address/0x32400084c286cf3e17e7b677ea9583e60a000324) contract. These transactions are recorded in the **"Internal Transactions"** section of Etherscan due to their internal nature.
:::

In summary, the **Transactions** section in Etherscan displays transactions between external addresses, while the **Internal Transactions** section displays transactions that occur within smart contracts.
```

> 
> ### How to check your internal transactions
> 
> To check the internal transactions of an Ethereum wallet, follow these steps:
> 
> 1. Launch your web browser and visit [Etherscan.io](https://etherscan.io/).
> 
> 2. In the search bar at the top of the page, enter the Ethereum wallet address you used to withdraw the funds. This should be the address where you withdrew the funds, not the destination address.
> 
> ![Etherscan search bar](../../assets/images/ether-search.png "Search for transactions")
> 
> 3. Once you've entered your wallet address, press the "Enter" key to search for your wallet on Etherscan.
> 
> 4. On the resulting page, scroll down to the **Internal Transactions** section. This section will display all of the internal transactions that have occurred within your wallet.
> 

Reviewer: ```suggestion

### How to check your internal transactions

To check the internal transactions of an Ethereum wallet, follow these steps:

1. Launch your web browser and visit [Etherscan.io](https://etherscan.io/).

2. In the search bar at the top of the page, enter the Ethereum wallet address you used to withdraw funds. This should be the address you withdrew funds from, not the destination address.

![Etherscan search bar](../../assets/images/ether-search.png "Search for transactions")

3. Click **Enter**.

4. Scroll down to the **Internal Transactions** section.

```

> Refer to the image below for more clarity.
> 
> ![Internal transaction](../../assets/images/internal-trx.png "Internal transaction in etherscan")
> 
> 5. Look for the internal transaction that corresponds to your withdrawal. You should see a transaction that shows the withdrawal amount being sent from the bridge.
> 
> ::: tip
> 
> Any withdrawal transactions from zkSync Era will appear as transactions from **zkSync Era: Diamond Proxy** to your wallet address.
> 
> :::
> 
> 6. Once you've located the transaction, you can click on the **Parent Tx Hash** to view more details about the transaction, including the block it was mined in, gas usage, and sender and recipient addresses.
> 
> ![Parent transaction hash](../../assets/images/parent-hash.png "Parent transaction hash in etherscan")
> 
> That's it! You should now be able to find your withdrawal transaction in the **Internal Transactions** section on Etherscan. If for any reason you're still unable to see your transaction after withdrawal please contact [support](../fundamentals/interacting.md#zksync-era-support) and we will look into it.

Reviewer: ```suggestion

![Internal transaction](../../assets/images/internal-trx.png "Internal transactions in Etherscan")

5. Look for the internal transaction that corresponds to your withdrawal. You should see a transaction that shows the withdrawal amount coming from the bridge.

::: tip
Withdrawal transactions from zkSync Era appear as transactions from **zkSync Era: Diamond Proxy** to your wallet address.
:::

6. Once you've located the transaction, click on the **Parent Tx Hash** to view more details about the transaction, including the block it was mined in, gas usage, and sender and recipient addresses.

![Parent transaction hash](../../assets/images/parent-hash.png "Parent transaction hash in etherscan")

If for any reason you're still unable to see your transaction after withdrawal and following these steps, please contact [support](../fundamentals/interacting.md#zksync-era-support) and we will look into it.
```
